### PR TITLE
feat(api): add /policies endpoint to list all registered policies

### PR DESCRIPTION
**Pull request by lahabana**
- Add an endpoint to see all policies registered with the cp.
This should make simpler and easier to have a dynamically generated api, gui...

Signed-off-by: Charly Molter <charly.molter@konghq.com>

### Checklist prior to review

- [x] Link to docs PR or issue -- https://github.com/kumahq/kuma-website/pull/935 
- [x] Link to UI issue or PR -- https://github.com/kumahq/kuma-gui/issues/336
- [x] Is the [issue worked on linked][1]? -- No issue
- [x] Unit Tests -- Yes
- [x] E2E Tests -- Yes
- [x] Manual Universal Tests -- Yes
- [x] Manual Kubernetes Tests -- Yes
- [x] Do you need to update [`UPGRADE.md`](/UPGRADE.md)? -- N/A
- [x] Does it need to be backported according to the [backporting policy](/CONTRIBUTING.md#backporting)? -- N/A

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: test/e2e_env/universal/universal_suite_test.go [hunk 1]
```diff
@@ -87,3 +88,4 @@ var _ = Describe("RateLimit", ratelimit.Policy, Ordered)
 var _ = Describe("ProxyTemplate", proxytemplate.ProxyTemplate, Ordered)
 var _ = Describe("Matching", matching.Matching, Ordered)
 var _ = Describe("Mtls", mtls.Policy, Ordered)
+var _ = FDescribe("Apis", api.Api, Ordered)
```
Comment (jakubdyszkiewicz) [inline test/e2e_env/universal/universal_suite_test.go]: F

File: pkg/api-server/policies_endpoints.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+		}
+		sort.SliceStable(response.Policies, func(i, j int) bool {
+			return response.Policies[i].Name < response.Policies[j].Name
+		})
+
+		if err := resp.WriteAsJson(response); err != nil {
+			log.Error(err, "Could not write the index response")
+		}
+	}))
+	return nil
```
Comment (jakubdyszkiewicz) [inline pkg/api-server/policies_endpoints.go:35]: why error then if it's always nil

File: pkg/api-server/policies_endpoints.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+func addPoliciesWsEndpoints(ws *restful.WebService, mode core.CpMode, readOnly bool, defs []model.ResourceTypeDescriptor) error {
+	ws.Route(ws.GET("/policies").To(func(req *restful.Request, resp *restful.Response) {
+		response := types.PoliciesResponse{}
+		for _, def := range defs {
+			if !def.IsPolicy {
+				continue
+			}
+			response.Policies = append(response.Policies, types.PolicyEntry{
+				Name:        string(def.Name),
+				ReadOnly:    readOnly || core.Zone == mode,
```
Comment (jakubdyszkiewicz) [inline pkg/api-server/policies_endpoints.go:22]: do we plan this to be different for different policies? if not, maybe this should be top level next to `policies`?

File: api/mesh/options.proto [hunk 4]
```diff
@@ -40,6 +40,9 @@ message KumaResourceOptions {
   // If resource has more than one version, then the flag defines which version
   // is used in the storage. All other versions must be convertible to it.
   bool storage_version = 14;
+
+  // Display Name
```
Comment (jakubdyszkiewicz) [inline api/mesh/options.proto:44]: That's not very descriptive. I liked that one
```
	// DisplayName the name of the policy showed as plural to be displayed in the UI and maybe CLI
```

Comment (michaelbeaumont) [reply api/mesh/options.proto:44]: IMO a sentence that just restates "this is the name for displaying" is more noisy than helpful. Not every field needs docs.

If we're gonna add the name of the field as the first word, it should be a proper sentence. But I blame godoc's ridiculous convention for that

File: pkg/api-server/testdata/policies_list.golden.json [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+   "displayName": "Fault Injections"
+  },
+  {
+   "name": "HealthCheck",
+   "readOnly": false,
+   "path": "health-checks",
+   "displayName": "Health Checks"
+  },
+  {
+   "name": "MeshGateway",
```
Comment (michaelbeaumont) [inline pkg/api-server/testdata/policies_list.golden.json:28]: I've said this before but we shouldn't call this a policy, it's required configuration for a type of `Dataplane`. If this is a policy, "policy" means nothing.

File: tools/resource-gen/genutils/util.go [hunk 6]
```diff
@@ -93,6 +97,16 @@ func ToResourceInfo(desc protoreflect.MessageDescriptor) ResourceInfo {
 			}
 		}
 	}
+	out.IsPolicy = !out.SkipRegistration && !out.Global && !out.WsAdminOnly && !out.WsReadOnly && out.ResourceType != "Dataplane"
```
Comment (michaelbeaumont) [inline tools/resource-gen/genutils/util.go:100]: Why is `Global` or `ReadOnly` not a policy?

File: pkg/api-server/policies_endpoints.go (hunk 3 above)
Comment (lahabana) [reply pkg/api-server/policies_endpoints.go:22]: I'm not sure I felt that this was just more flexible for the future (likely if we deprecate things for example). ATM the whole policy engine enables this.

Comment (lahabana) [reply pkg/api-server/policies_endpoints.go:22]: Actually adding `def.ReadOnly` too :)

File: pkg/api-server/testdata/policies_list.golden.json (hunk 5 above)
Comment (lahabana) [reply pkg/api-server/testdata/policies_list.golden.json:28]: I think both MeshGateway and ExternalServices are weird. I'm not sure it's the scope of this issue to fix this.

TBH I'm not sure what you define as a Policy. To me it's things you apply to a set of Dps to alter the conf. In this case MeshGateway fits this def

File: tools/resource-gen/genutils/util.go (hunk 6 above)
Comment (lahabana) [reply tools/resource-gen/genutils/util.go:100]: Cause all policies depend of a Mesh. ReadOnly is only used for internal purpose

Comment (lahabana) [reply tools/resource-gen/genutils/util.go:100]: Adding a comment for this

File: pkg/api-server/testdata/policies_list.golden.json (hunk 5 above)
Comment (lahabana) [reply pkg/api-server/testdata/policies_list.golden.json:28]: Maybe I'm unsure. Feels like this is a team conversation to have here. In any case this is not the scope of this change (this only bases itself on the existing things we define as "a policy")

Comment (michaelbeaumont) [reply pkg/api-server/testdata/policies_list.golden.json:28]: I guess

**Review by jakubdyszkiewicz — APPROVED** [review]

**Pull request merged**